_posts/2018-11-01-Inter-service-communication.md

```diff
@@ -0,0 +1,327 @@
+* `fallback`: If Hystrix is enabled, you can implement a fallback method.  
+
+{% highlight java %}
+    @Component
+    public class FallbackAuthClient implements AuthClient {
+        @Override
+        public boolean isTokenValid(@RequestHeader("Authorization") String token) {
+            return false;
+        }
+    }
+{% endhighlight %}
+
+* `configuration`: is for extra configuration like logging, interceptors, etc... more on that below.
+* `@RequestMapping`: Spring Cloud adds support for Spring MVC annotations and for using the same `HttpMessageConverters` used by default in Spring Web.
+
+## OpenFeign
+To create an OpenFeign client we need an interface and a Feign builder that tells the interface it is an OpenFeign client.
+
+{% highlight java %}
+
+public interface AuthClient {
+    @RequestLine("GET /auth")
+    @Headers({"Authorization: {token}", "Accept: application/hal+json"})
+    boolean isValid(@Param("token") String token);
+
+{% endhighlight %}
+* `@RequestLine`: is defining which verb and which uri path you are communicating to. 
+* `@Headers`: is defining the request headers that come with the request.
+
+# The builder
+OpenFeign provides us with a builder like pattern for our clients.
+When we want to customize, we just add our own customization to the builder. 
+To see the builder at work, let's create a bean of our client and return a Feign builder.
+It's important to let the builder know which interface he has to target for communication. 
+The second parameter is most likely the base url where all the requests begin. 
+Get your urls from the yml or properties file with the help of `@Value`.
+
+{% highlight java %}
+
+   @Value("${base.url}")
+   private String baseServerUrl;
+    
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+# Different kind of clients
+The default HTTP client of OpenFeign uses `HttpUrlConnection` to execute its HTTP requests.
+You can configure another client (`ApacheHttpClient`, `OkHttpClient`, ...) as follows:
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .client(new ApacheHttpClient())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+## OkHttpClient
+OkHttp is an HTTP client that’s efficient by default:
+
+* HTTP/2 support allows all requests to the same host to share a socket.
+* Connection pooling reduces request latency (if HTTP/2 isn’t available).
+* Transparent GZIP shrinks download sizes.
+* Response caching avoids the network completely for repeat requests.
+
+## ApacheHttpClient
+The advantage of using `ApacheHttpClient` over the default client is that `ApacheHttpClient` sends more headers with the request, eg. `Content-Length`, which some servers expect.
+
+### Building a Mutual SSL client with Apache
```

Comment:
Seeing as one of the ToC items links to this item, I assume that this needs to be an `h1`?
This will also properly set the view after clicking the ToC item as it currently is a bit offset.
```suggestion
# Building a Mutual SSL client with Apache
```